lib/vanagon/driver.rb

```diff
@@ -13,26 +13,38 @@ class Driver
     attr_accessor :platform, :project, :target, :workdir, :verbose, :preserve
     attr_accessor :timeout, :retry_count
 
-    def initialize(platform, project, options = { :configdir => nil, :target => nil, :engine => nil, :components => nil, :skipcheck => false, :verbose => false, :preserve => false }) # rubocop:disable Metrics/AbcSize
+    def initialize(platform, project, options = { :configdir => nil, :target => nil, :engine => nil, :components => nil, :skipcheck => false, :verbose => false, :preserve => false, :only_build => nil }) # rubocop:disable Metrics/AbcSize
       @verbose = options[:verbose]
       @preserve = options[:preserve]
 
       @@configdir = options[:configdir] || File.join(Dir.pwd, "configs")
       components = options[:components] || []
+      only_build = options[:only_build]
       target = options[:target]
       engine = options[:engine] || 'pooler'
 
       @platform = Vanagon::Platform.load_platform(platform, File.join(@@configdir, "platforms"))
       @project = Vanagon::Project.load_project(project, File.join(@@configdir, "projects"), @platform, components)
       @project.settings[:verbose] = options[:verbose]
       @project.settings[:skipcheck] = options[:skipcheck]
+      filter_out_components(only_build) if only_build
       loginit('vanagon_hosts.log')
 
       load_engine(engine, @platform, target)
     rescue LoadError => e
       raise Vanagon::Error.wrap(e, "Could not load the desired engine '#{engine}'")
     end
 
+    def filter_out_components(only_build)
+      # map each element in the only_build array to it's set of filtered components, then
+      # flatten all the results in to one array and set project.components to that.
+      @project.components = only_build.flat_map { |comp| @project.filter_component(comp) }.uniq
```

Comment:
![bad-smell](https://cloud.githubusercontent.com/assets/344926/20733022/55a6b708-b646-11e6-9446-c70e3c6c3f67.jpg)

(not a blocker; just... whoof, any time you're running `#uniq` after `#map`, you know you should probably fix something further up in the code path.